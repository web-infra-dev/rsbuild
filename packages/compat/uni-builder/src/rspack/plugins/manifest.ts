import type { RsbuildPlugin, RsbuildPluginAPI } from '@rsbuild/core';
import { generateManifest } from '../../shared/manifest';

export const pluginManifest = (): RsbuildPlugin<RsbuildPluginAPI> => ({
  name: 'plugin-manifest',

  setup(api) {
    api.modifyBundlerChain(async (chain, { CHAIN_ID }) => {
      const { WebpackManifestPlugin } = await import('rspack-manifest-plugin');
      const publicPath = chain.output.get('publicPath');

      chain.plugin(CHAIN_ID.PLUGIN.MANIFEST).use(WebpackManifestPlugin, [
        {
          fileName: 'asset-manifest.json',
          publicPath,
          generate: generateManifest,
        },
      ]);
    });
  },
});
