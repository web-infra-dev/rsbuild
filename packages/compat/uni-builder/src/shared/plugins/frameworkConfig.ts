import { fse, type DefaultRsbuildPlugin } from '@rsbuild/shared';

export const pluginFrameworkConfig = (
  configPath: string,
): DefaultRsbuildPlugin => ({
  name: 'plugin-framework-config',

  setup(api) {
    api.modifyBundlerChain((chain) => {
      if (!fse.existsSync(configPath)) {
        return;
      }

      const cache = chain.get('cache');

      cache.buildDependencies = {
        ...cache.buildDependencies,
        frameworkConfig: [configPath],
      };

      chain.cache(cache);
    });
  },
});
