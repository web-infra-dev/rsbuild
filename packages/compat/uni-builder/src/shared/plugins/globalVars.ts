import {
  mergeChainedOptions,
  type Define,
  type DefaultRsbuildPlugin,
} from '@rsbuild/shared';
import type { ChainedGlobalVars } from '../../types';

export const pluginGlobalVars = (
  options?: ChainedGlobalVars,
): DefaultRsbuildPlugin => ({
  name: 'plugin-global-vars',

  setup(api) {
    api.modifyBundlerChain((chain, { env, target, bundler }) => {
      if (!options) {
        return;
      }

      const globalVars = mergeChainedOptions({
        defaults: {},
        options,
        utils: { env, target },
      });

      const serializedVars: Define = {};

      Object.entries(globalVars).forEach(([key, value]) => {
        serializedVars[key] = JSON.stringify(value) ?? 'undefined';
      });

      chain.plugin('globalVars').use(bundler.DefinePlugin, [serializedVars]);
    });
  },
});
