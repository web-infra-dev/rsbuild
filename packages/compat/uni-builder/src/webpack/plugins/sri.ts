import { SubresourceIntegrityPlugin } from 'webpack-subresource-integrity';
import type { RsbuildPlugin } from '@rsbuild/webpack';
import type { SriOptions } from '../../types';

export const pluginSRI = (options: SriOptions | boolean): RsbuildPlugin => ({
  name: 'plugin-sri',

  setup(api) {
    api.modifyWebpackChain((chain, { CHAIN_ID }) => {
      chain.output.crossOriginLoading('anonymous');
      chain
        .plugin(CHAIN_ID.PLUGIN.SUBRESOURCE_INTEGRITY)
        .use(SubresourceIntegrityPlugin, [
          typeof options === 'object' ? options : undefined,
        ]);
    });
  },
});
