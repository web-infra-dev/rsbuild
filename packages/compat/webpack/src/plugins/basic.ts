import path from 'path';
import type { RsbuildPlugin } from '../types';
import { applyBasicPlugin } from '@rsbuild/shared';

/**
 * Provide some basic configs of webpack
 */
export const pluginBasic = (): RsbuildPlugin => ({
  name: 'plugin-basic',

  setup(api) {
    applyBasicPlugin(api);

    api.modifyWebpackChain(async (chain, { env }) => {
      // Disable webpack performance hints.
      // These logs are too complex
      chain.performance.hints(false);

      // This will be futureDefaults in webpack 6
      chain.module.parser.merge({
        javascript: {
          exportsPresence: 'error',
        },
      });

      if (env === 'development') {
        // Set correct path for source map
        // this helps VS Code break points working correctly in monorepo
        chain.output.devtoolModuleFilenameTemplate(
          (info: { absoluteResourcePath: string }) =>
            path.resolve(info.absoluteResourcePath).replace(/\\/g, '/'),
        );
      }
    });
  },
});
