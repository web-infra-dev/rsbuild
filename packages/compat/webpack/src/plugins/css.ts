import {
  CSS_REGEX,
  resolvePackage,
  isUseCssExtract,
  getPostcssConfig,
  ModifyChainUtils,
  isUseCssSourceMap,
  mergeChainedOptions,
  getCssLoaderOptions,
  getBrowserslistWithDefault,
  getCssModuleLocalIdentName,
  type Context,
  type BundlerChainRule,
} from '@rsbuild/shared';
import type {
  RsbuildPlugin,
  CSSExtractOptions,
  NormalizedConfig,
} from '../types';

export async function applyBaseCSSRule({
  rule,
  config,
  context,
  utils: { target, isProd, isServer, CHAIN_ID, isWebWorker, getCompiledPath },
  importLoaders = 1,
}: {
  rule: BundlerChainRule;
  config: NormalizedConfig;
  context: Context;
  utils: ModifyChainUtils;
  importLoaders?: number;
}) {
  const browserslist = await getBrowserslistWithDefault(
    context.rootPath,
    config,
    target,
  );

  // 1. Check user config
  const enableExtractCSS = isUseCssExtract(config, target);
  const enableSourceMap = isUseCssSourceMap(config);
  const enableCSSModuleTS = Boolean(config.output.enableCssModuleTSDeclaration);

  // 2. Prepare loader options
  const localIdentName = getCssModuleLocalIdentName(config, isProd);

  const cssLoaderOptions = getCssLoaderOptions({
    config,
    enableSourceMap,
    importLoaders,
    isServer,
    isWebWorker,
    localIdentName,
  });

  // 3. Create webpack rule
  // Order: style-loader/mini-css-extract -> css-loader -> postcss-loader
  if (!isServer && !isWebWorker) {
    // use mini-css-extract-plugin loader
    if (enableExtractCSS) {
      const extraCSSOptions: Required<CSSExtractOptions> =
        typeof config.tools.cssExtract === 'object'
          ? config.tools.cssExtract
          : {
              loaderOptions: {},
              pluginOptions: {},
            };

      rule
        .use(CHAIN_ID.USE.MINI_CSS_EXTRACT)
        .loader(require('mini-css-extract-plugin').loader)
        .options(extraCSSOptions.loaderOptions)
        .end();
    }
    // use style-loader
    else {
      const styleLoaderOptions = mergeChainedOptions({
        defaults: {},
        options: config.tools.styleLoader,
      });

      rule
        .use(CHAIN_ID.USE.STYLE)
        .loader(require.resolve('style-loader'))
        .options(styleLoaderOptions)
        .end();
    }

    // use css-modules-typescript-loader
    if (enableCSSModuleTS && cssLoaderOptions.modules) {
      rule
        .use(CHAIN_ID.USE.CSS_MODULES_TS)
        .loader(
          resolvePackage(
            '@rsbuild/shared/css-modules-typescript-loader',
            __dirname,
          ),
        )
        .options({
          modules: cssLoaderOptions.modules,
        })
        .end();
    }
  } else {
    rule
      .use(CHAIN_ID.USE.IGNORE_CSS)
      .loader(resolvePackage('@rsbuild/shared/ignore-css-loader', __dirname))
      .end();
  }

  rule
    .use(CHAIN_ID.USE.CSS)
    .loader(getCompiledPath('css-loader'))
    .options(cssLoaderOptions)
    .end();

  if (!isServer && !isWebWorker) {
    const postcssLoaderOptions = getPostcssConfig({
      enableSourceMap,
      browserslist,
      config,
    });

    rule
      .use(CHAIN_ID.USE.POSTCSS)
      .loader(getCompiledPath('postcss-loader'))
      .options(postcssLoaderOptions)
      .end();
  }

  // CSS imports should always be treated as sideEffects
  rule.merge({ sideEffects: true });

  // Enable preferRelative by default, which is consistent with the default behavior of css-loader
  // see: https://github.com/webpack-contrib/css-loader/blob/579fc13/src/plugins/postcss-import-parser.js#L234
  rule.resolve.preferRelative(true);
}

export const pluginCss = (): RsbuildPlugin => {
  return {
    name: 'plugin-css',
    setup(api) {
      api.modifyBundlerChain(async (chain, utils) => {
        const rule = chain.module.rule(utils.CHAIN_ID.RULE.CSS);
        const config = api.getNormalizedConfig();
        rule.test(CSS_REGEX);
        await applyBaseCSSRule({
          rule,
          utils,
          config,
          context: api.context,
        });
      });
    },
  };
};
