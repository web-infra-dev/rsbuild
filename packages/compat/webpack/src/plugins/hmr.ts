import { isUsingHMR, type RsbuildPlugin } from '@rsbuild/shared';

export const pluginHMR = (): RsbuildPlugin => ({
  name: 'rsbuild-webpack:hmr',

  setup(api) {
    api.modifyBundlerChain((chain, utils) => {
      const config = api.getNormalizedConfig();

      if (!isUsingHMR(config, utils)) {
        return;
      }

      const { bundler, CHAIN_ID } = utils;
      chain.plugin(CHAIN_ID.PLUGIN.HMR).use(bundler.HotModuleReplacementPlugin);
    });
  },
});
