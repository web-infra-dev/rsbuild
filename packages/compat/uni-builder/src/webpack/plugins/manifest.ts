import type { RsbuildPlugin } from '@rsbuild/webpack';
import { generateManifest } from '../../shared/manifest';

export const pluginManifest = (): RsbuildPlugin => ({
  name: 'uni-builder:manifest',

  setup(api) {
    api.modifyWebpackChain(async (chain, { CHAIN_ID }) => {
      const { WebpackManifestPlugin } = await import('webpack-manifest-plugin');
      const publicPath = chain.output.get('publicPath');

      chain.plugin(CHAIN_ID.PLUGIN.MANIFEST).use(WebpackManifestPlugin, [
        {
          fileName: 'asset-manifest.json',
          publicPath,
          generate: generateManifest,
        },
      ]);
    });
  },
});
