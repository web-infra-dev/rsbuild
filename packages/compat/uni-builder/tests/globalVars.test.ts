import { createStubRsbuild } from '@rsbuild/test-helper';
import { pluginGlobalVars } from '../src/shared/plugins/globalVars';

describe('plugin-global-vars', () => {
  const cases = [
    {
      name: 'globalVars',
      options: {
        'process.env.foo': 'foo',
        'import.meta.bar': {
          a: 'bar',
          b: false,
          c: { d: 42 },
        },
        'window.baz': [null, 'baz'],
      },
    },
    {
      name: 'globalVars function',
      options: (obj: any, { env, target }: any) => {
        obj.ENV = env;
        obj.TARGET = target;
      },
    },
  ];

  it.each(cases)('$name', async (item) => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginGlobalVars(item.options)],
    });

    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });
});
