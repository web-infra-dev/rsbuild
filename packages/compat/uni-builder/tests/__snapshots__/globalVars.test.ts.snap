// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin-global-vars > 'globalVars function' 1`] = `
{
  "plugins": [
    DefinePlugin {
      "_options": {
        "ENV": ""test"",
        "TARGET": ""web"",
      },
      "name": "DefinePlugin",
    },
  ],
}
`;

exports[`plugin-global-vars > 'globalVars' 1`] = `
{
  "plugins": [
    DefinePlugin {
      "_options": {
        "import.meta.bar": "{"a":"bar","b":false,"c":{"d":42}}",
        "process.env.foo": ""foo"",
        "window.baz": "[null,"baz"]",
      },
      "name": "DefinePlugin",
    },
  ],
}
`;
